File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result =  [0] * n
        result[0] = 1
        
        # result[i] = product of all numbers to the left
        for i in range(1, n):
            result[i] = result[i - 1] * nums[i - 1];
    
        # result[i] *= product of all numbers to the rigth
        right = 1
        for i in range(n - 1, -1, -1):
            result[i] *= right
            right *= nums[i]
        
        return result

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        int[] leftP = new int[n];
        int[] rightP = new int[n];

        leftP[0] = 1;
        rightP[n - 1] = 1;

        for (int i = 1; i < n; i++) {
            leftP[i] = leftP[i - 1] * nums[i - 1];
            rightP[n - 1 - i] = rightP[n - i] * nums[n - i];
        }

        for (int i = 0; i < n; i++) {
            result[i] = leftP[i] * rightP[i];
        }


        return result;
    }
}